### PR TITLE
Modularize Phase1 into post-Merge features: Sharding, Custody Game, Data Availability Sampling

### PR DESCRIPTION
**Pull request by protolambda**
This takes the older phase1 PRs #2146 and #2172 and rebases it onto dev (including the new Merge)

Progress:
- [x] Rebase DAS PR by @vbuterin onto latest The-Merge work
- [x] Rebase out clutter (generated files and such) that got into the old DAS PR
- [x] Clean up old shard code, remove ShardTransition accounting in favor of independent shard blobs.
- [x] Remove legacy light client design in favor of new Altair design.
- [x] Split custody game, DAS, and base sharding docs
- [x] Remove old phase1 pyspec building
- [x] Refactor custody game logic out of sharding, implement as extension instead.
- ~~Update block processing and epoch processing to call Altair and Merge-fork processing functions, not just phase0~~
  - blocked by merge work, ~~waiting for that branch to adopt Altair changes.~~ rebase of Merge in Altair in future PRs, use phase0 for now.
- [x] Update BeaconState type in sharding spec to support Merge.
- [ ] Update DAS spec to implement discv5.1 based long-range sample method (for Future PR maybe?)
- [x] Clean up old phase1 references in testing
- [x] Retain proof of custody testing (no proof of custody executable spec, tests are being skipped for now)
- [x] Ratain start-shard unit tests (no sharding executable spec, tests are being skipped for now)
- [x] Drop shard-transition tests in favor of new data-availability focused shards (no strict transitions, just data linked into beacon chain)
- [x] Update block processing and epoch processing helpers with missing merge/sharding/proof-of-custody methods, dropping shard-transition related processing.
- [x] Add phase naming for `SHARDING`, `CUSTODY_GAME`, `DAS` to tag tests with (to be enabled later). Note: custody tests need more in-depth updates, as the sharding design changed.
- [x] Update Makefile to drop older phase1 work
- [x] Drop old phase1 configs, split into new feature configs.
- [x] Update README

The bigger picture of this PR is to accelerate sharding after the merge, and enable early testnets to have shard-data, while DAS and proof of custody continue to be developed. Features, not phases.

### CODE REVIEW

**Comment by 304s304** [comment]
Run test

**Comment by hwwhww** [comment]
And:
- [ ] Fix test generators. There are some `from eth2spec.test.context import PHASE1` in testgen `main.py`s.

🙏

File: specs/sharding/beacon-chain.md [hunk 1]
```diff
@@ -0,0 +1,690 @@
...
+class AttestationData(Container):
+    slot: Slot
+    index: CommitteeIndex
+    # LMD GHOST vote
+    beacon_block_root: Root
+    # FFG vote
+    source: Checkpoint
+    target: Checkpoint
+    # Shard header root
+    shard_header_root: Root
```
Comment (djrtwo) [inline specs/sharding/beacon-chain.md:150]:
```suggestion
    shard_header_root: Root  # [New in Sharding]
```

File: specs/sharding/beacon-chain.md [hunk 2]
```diff
@@ -0,0 +1,690 @@
...
+```python
+def next_power_of_two(x):
+    return 2 ** ((x - 1).bit_length())
+```
+
+#### `compute_previous_slot`
+
+```python
+def compute_previous_slot(slot: Slot) -> Slot:
+    if slot > 0:
```
Comment (djrtwo) [inline specs/sharding/beacon-chain.md:245]: Should `0` be replaced with `GENESIS_SLOT` in this function?

File: specs/sharding/beacon-chain.md [hunk 3]
```diff
@@ -0,0 +1,690 @@
...
+
+
+### Block processing
+
+```python
+def process_block(state: BeaconState, block: BeaconBlock) -> None:
+    process_block_header(state, block)
+    process_randao(state, block.body)
+    process_eth1_data(state, block.body)
+    process_operations(state, block.body)  # [Modified]
```
Comment (djrtwo) [inline specs/sharding/beacon-chain.md:422]:
```suggestion
    process_operations(state, block.body)  # [Modified in Sharding]
```

File: specs/sharding/beacon-chain.md [hunk 4]
```diff
@@ -0,0 +1,690 @@
...
+    process_application_payload(state, block.body)  # [Part of the Merge]
```
Comment (djrtwo) [inline specs/sharding/beacon-chain.md:423]:
```suggestion
    process_application_payload(state, block.body)  # [New in Merge]
```

File: specs/sharding/beacon-chain.md [hunk 5]
```diff
@@ -0,0 +1,690 @@
...
+def process_epoch(state: BeaconState) -> None:
+    process_justification_and_finalization(state)
+    process_rewards_and_penalties(state)
+    process_registry_updates(state)
+
+    process_slashings(state)
+
+    # Sharding
+    process_pending_headers(state)
+    charge_confirmed_header_fees(state)
```
Comment (djrtwo) [inline specs/sharding/beacon-chain.md:562]: suggest "process_" prefix
```suggestion
    process_confirmed_header_fees(state)
```

File: specs/sharding/beacon-chain.md [hunk 6]
```diff
@@ -0,0 +1,690 @@
...
+    process_historical_roots_update(state)
+    process_participation_record_updates(state)
+
+    process_shard_epoch_increment(state)
+```
+
+#### Pending headers
+
+```python
+
```
Comment (djrtwo) [inline specs/sharding/beacon-chain.md:580]:
```suggestion
```

File: specs/sharding/beacon-chain.md [hunk 7]
```diff
@@ -0,0 +1,690 @@
...
+```python
+def reset_pending_headers(state: BeaconState) -> None:
+    state.previous_epoch_pending_shard_headers = state.current_epoch_pending_shard_headers
+    state.current_epoch_pending_shard_headers = []
+    # Add dummy "empty" PendingShardHeader (default vote for if no shard header available)
+    next_epoch = get_current_epoch(state) + 1
+    next_epoch_start_slot = compute_start_slot_at_epoch(next_epoch)
+    for slot in range(next_epoch_start_slot, next_epoch_start_slot + SLOTS_IN_EPOCH):
+        for index in range(get_committee_count_per_slot(next_epoch)
```
Comment (djrtwo) [inline specs/sharding/beacon-chain.md:671]:
```suggestion
        for index in range(get_committee_count_per_slot(next_epoch):
```

File: tests/core/pyspec/eth2spec/test/context.py [hunk 8]
```diff
@@ -29,10 +27,15 @@ def reload_specs():
...
 ALTAIR = SpecForkName('altair')
 
-ALL_PHASES = (PHASE0, PHASE1, ALTAIR)
+# Experimental phases (not included in default "ALL_PHASES"):
+MERGE = SpecForkName('merge')
+SHARDING = SpecForkName('sharding')
+CUSTODY_GAME = SpecForkName('custody_game')
+DAS = SpecForkName('das')
+
+ALL_PHASES = (PHASE0, ALTAIR)
```
Comment (djrtwo) [inline tests/core/pyspec/eth2spec/test/context.py:38]: Doesn't MERGE pass all PHASE0 tests? I think we should include it in CI

**Review by djrtwo — COMMENTED** [review]
lots of minor formatting and link fixes

File: specs/sharding/beacon-chain.md (hunk 2 above)
Comment (protolambda) [reply specs/sharding/beacon-chain.md:245]: It is `Slot(0)` below, and for the underflow problem more so than configuration (`GENESIS_SLOT` is pretty much a constant at this point)

File: Makefile [hunk 9]
```diff
@@ -112,10 +116,11 @@ check_toc: $(MARKDOWN_FILES:=.toc)
 codespell:
 	codespell . --skip ./.git -I .codespell-whitelist
 
+# TODO: add future merge, sharding, etc. packages to linting.
 lint: pyspec
 	. venv/bin/activate; cd $(PY_SPEC_DIR); \
 	flake8  --config $(LINTER_CONFIG_FILE) ./eth2spec \
-	&& mypy --config-file $(LINTER_CONFIG_FILE) -p eth2spec.phase0 -p eth2spec.phase1 -p eth2spec.altair
+	&& mypy --config-file $(LINTER_CONFIG_FILE) -p eth2spec.phase0 -p eth2spec.altair
```
[the remark this answers is not in the document]
Comment (protolambda) [reply Makefile:123]: We can lint it when it becomes executable, in a future PR.

File: tests/core/pyspec/eth2spec/test/context.py (hunk 8 above)
Comment (protolambda) [reply tests/core/pyspec/eth2spec/test/context.py:38]: Let's do so in another PR before this one grows too large

File: specs/custody_game/custody-game.md [hunk 10]
```diff
@@ -4,6 +4,7 @@
```
[the remark this answers is not in the document]
Comment (protolambda) [reply specs/custody_game/custody-game.md]: Renamed, also updated the references in adjacent doc and readme.

**Review by djrtwo — APPROVED** [review]

**Pull request merged**